server/src/main/java/org/elasticsearch/cluster/ClusterState.java

```diff
@@ -777,13 +777,35 @@ public DiscoveryNodes nodes() {
             return nodes;
         }
 
+        // Deprecate to keep downstream projects compiling
+        @Deprecated(forRemoval = true)
         public Builder putTransportVersion(String nodeId, TransportVersion transportVersion) {
-            // TODO[wrb]: system index mappings versions will be added in a followup
-            compatibilityVersions.put(nodeId, new CompatibilityVersions(Objects.requireNonNull(transportVersion, nodeId), Map.of()));
+            return putCompatibilityVersions(nodeId, transportVersion, Map.of());
+        }
+
+        public Builder putCompatibilityVersions(
+            String nodeId,
+            TransportVersion transportVersion,
+            Map<String, SystemIndexDescriptor.MappingsVersion> systemIndexMappingsVersions
+        ) {
+            return putCompatibilityVersions(
+                nodeId,
+                new CompatibilityVersions(Objects.requireNonNull(transportVersion, nodeId), systemIndexMappingsVersions)
+            );
+        }
+
+        public Builder putCompatibilityVersions(String nodeId, CompatibilityVersions versions) {
+            compatibilityVersions.put(nodeId, versions);
             return this;
         }
 
+        // Deprecate to keep downstream projects compiling
+        @Deprecated(forRemoval = true)
         public Builder compatibilityVersions(Map<String, CompatibilityVersions> versions) {
+            return compatibilityVersionsMap(versions);
+        }
+
+        public Builder compatibilityVersionsMap(Map<String, CompatibilityVersions> versions) {
```

Comment:
Do we need to rename this method?

Reply:
How about `nodeIdsToCompatibilityVersions`?

Reply:
Works for me!